### PR TITLE
feat: improve build speed

### PR DESCRIPTION
**Pull request by PranavPurwar**
By stopping the unnecessary merging of dex files during debug build

### CODE REVIEW

**Comment by PranavPurwar** [comment]
oh wait before merging it, i think it should check min sdk version before skipping dex merging too

**Comment by PranavPurwar** [comment]
ok completed now

**Comment by khaled-0** [comment]
I don't get it.
What is your change supposed to do?
Did you miss anything 🤔

**Comment by PranavPurwar** [comment]
merging dex files of libraries isn't necessary when minimum sdk is above 21 (due to multidex support) and a debug build is being done, so it will just skip that step.

**Comment by tyron12233** [comment]
While its true that do not need to merge the dex files for each library on a debug build, you should still add each dex files on the apk. If you don't add them then the generated apk will crash as it can't find the classes of the libraries

**Comment by PranavPurwar** [comment]
ok done, anything else ?

**Comment by JavkhlanK** [comment]
testing out proposed changes will always be a great idea, btw

**Comment by khaled-0** [comment]
> testing out proposed changes will always be a great idea, btw

Fine, I'll do it myself.
Snap

File: app/src/main/java/a/a/a/Dp.java [hunk 1]
```diff
@@ -169,7 +168,24 @@ public void a(iI iIVar, String str) {
      * @throws Exception Thrown if dexing had problems
      */
     public final void a(String outputPath, ArrayList<String> dexes) throws Exception {
-        dexLibraries(outputPath, dexes);
+        try {
+            if (!Integer.parseInt(this.a) >= 21) {
```
Comment (khaled-0) [inline app/src/main/java/a/a/a/Dp.java:172]: image
! cant be directly applied sir
surround when like picture

Comment (tyron12233) [reply app/src/main/java/a/a/a/Dp.java:172]: or just invert the operator

File: app/src/main/java/a/a/a/Dp.java [hunk 2]
```diff
@@ -169,7 +168,24 @@ public void a(iI iIVar, String str) {
...
+                dexLibraries(outputPath, dexes);
+            } else {
+                Iterator it = dexes.iterator();
+                while (it.hasNext()) {
+                    String dexPath = it.next();
```
Comment (khaled-0) [inline app/src/main/java/a/a/a/Dp.java:177]: image
how to fix this sir

Comment (tyron12233) [reply app/src/main/java/a/a/a/Dp.java:177]: use the generic type `Iterator<Spring>`

**Comment by khaled-0** [comment]
sir i get error when merging dex 
``No direct method <init>(La/a/a/Dp;Ljava/io/File;ZLcom/besome/sketch/design/DesignActivity$a;)V in class Lmod/jbk/build/compiler/resource/ResourceCompiler; or its super classes (declaration of 'mod.jbk.build.compiler.resource.ResourceCompiler' appears in /data/app/~~Fyo07E1fOYNy_lvKMeWe3Q==/com.sketchware.remod-NS2SwT9gDw2hSySOcptxJQ==/base.apk!classes6.dex)``

**Comment by PranavPurwar** [comment]
that appears to be related to ResourceCompiler

**Comment by PranavPurwar** [comment]
did you edit that file ? because that method is already present in the class

**Comment by khaled-0** [comment]
> did you edit that file ? because that method is already present in the class

i just made these changes on beta 6.4

**Comment by PranavPurwar** [comment]
that seems sussy, maybe a clean rebuild would solve that

**Comment by khaled-0** [comment]
congratulation sir ur pr works
With:
image

without
image

**Comment by PranavPurwar** [comment]
ok

**Comment by PranavPurwar** [comment]
send me the built apk so i run more tests with it

**Comment by PranavPurwar** [comment]
the first thing i would test: compiling a project with all appcompat, firebase, admob, map sdk, and all my local libs :trollface:

**Comment by khaled-0** [comment]
> send me the built apk so i run more tests with it

https://we.tl/t-mpKeiqgnUu
Built on top of latest beta, nothing extra.
Download link lasts for 10hrs maybe. Download fast

**Comment by PranavPurwar** [comment]
ok, thanks

**Comment by PranavPurwar** [comment]
what, it still merged dex files 😔 
Screenshot_2021-11-23-22-04-55

**Comment by khaled-0** [comment]
> what, it still merged dex files 😔 Screenshot_2021-11-23-22-04-55

I don't know sir

**Comment by PranavPurwar** [comment]
but sir the output dex file is has merged libs 😔

**Comment by khaled-0** [comment]
> but sir the output dex file is has merged libs 😔

Screenshot_20211123-230852
Sir the app also contains your code sir
And theres no other Dp.smali either.
This onle lives inside smali_classes6

**Comment by PranavPurwar** [comment]
ok sir, maybe i did something wrong in minApi thingy

**Comment by JavkhlanK** [comment]
`Dp#a` will never change btw

**Comment by JavkhlanK** [comment]
You'd wanna use ProjectSettings (`Dp#settings` in this case) instead

**Comment by PranavPurwar** [comment]
ok

**Comment by khaled-0** [comment]
> You'd wanna use ProjectSettings (`Dp#settings` in this case) instead

But, Dp#a is still 21 by default so his code should work🤔
Mine built faster

**Comment by khaled-0** [comment]
Screenshot_20211123-231833
Multiple dex sir.
And classes2.dex contains some of lubrary and project classed toi

File: app/src/main/java/a/a/a/Dp.java [hunk 3]
```diff
@@ -169,7 +169,8 @@ public void a(iI iIVar, String str) {
      */
     public final void a(String outputPath, ArrayList<String> dexes) throws Exception {
         try {
-            if (!(Integer.parseInt(this.a) >= 21)) {
+            int minSdkVersion = Integer.parseInt(settings.getValue(ProjectSettings.SETTING_MINIMUM_SDK_VERSION, "21"));
```
Comment (khaled-0) [inline app/src/main/java/a/a/a/Dp.java:172]: Use that upper Variable instead of hardcoaded "21"

**Comment by franciscowilton** [comment]
> > envie-me o apk integrado para que eu execute mais testes com ele
> 
> https://we.tl/t-mpKeiqgnUu Construído sobre a versão beta mais recente, nada extra. O link para download dura talvez 10 horas. Baixe rápido



> > send me the built apk so i run more tests with it
> 
> https://we.tl/t-mpKeiqgnUu Built on top of latest beta, nothing extra. Download link lasts for 10hrs maybe. Download fast

When compiling using Proguard the following error is returned: Compilation failed to complete, origin: /storage/emulated/0/.sketchware/mysc/601/bin/classes_proguard.jar

**Comment by PranavPurwar** [comment]
bruh, first of all, this is not a release
secondly, i tried it and got no error, probably you're doing something wrong @franciscowilton

File: app/src/main/java/a/a/a/Dp.java (hunk 3 above)
Comment (PranavPurwar) [reply app/src/main/java/a/a/a/Dp.java:172]: im not sure if that variable is not initialized in future, so i's rather use that hardcoded int

**Comment by franciscowilton** [comment]
> bruh, em primeiro lugar, isso não é um release em segundo lugar, tentei e não achei nenhum erro, provavelmente você está fazendo algo errado @franciscowilton

Yes I know it's not a release, I'm just testing.

**Comment by JavkhlanK** [comment]
I just noticed that the changed method never gets called anymore lol
image

**Comment by PranavPurwar** [comment]
lamo, so that's why that change didn't work

**Comment by PranavPurwar** [comment]
> congratulation sir ur pr works With: image
> 
> without image

don't get how that worked lmao

**Comment by khaled-0** [comment]
> > congratulation sir ur pr works With: image
> > without image
> 
> don't get how that worked lmao

:thenking:

**Comment by JavkhlanK** [comment]
I don't think it's a good idea to copy the DEX files to Internal storage, as that's redundant. Adding them directly to the APK would work too. The current changes are also not affecting the DEX merging step, so these changes aren't worth a merge, sorry.

**Comment by PranavPurwar** [comment]
ok

**Pull request closed without merging**